pyteal/ast/multi.py

```diff
@@ -0,0 +1,85 @@
+from typing import Callable, List, Union, TYPE_CHECKING
+
+from ..types import TealType
+from ..ir import TealOp, Op, TealBlock
+from .expr import Expr
+from .leafexpr import LeafExpr
+from .scratch import ScratchSlot
+
+if TYPE_CHECKING:
+    from ..compiler import CompileOptions
+
+
+class MultiValue(LeafExpr):
+    """Represents an operation that returns more than one value"""
+
+    def __init__(
+        self,
+        op: Op,
+        types: List[TealType],
+        *,
+        immediate_args: List[Union[int, str]] = None,
+        args: List[Expr] = None
+    ):
+        """Create a new MultiValue.
+
+        Args:
+            op: The operation that returns values.
+            types: The types of the returned values.
+            immediate_args (optional): Immediate arguments for the op. Defaults to None.
+            args (optional): Stack arguments for the op. Defaults to None.
+        """
+        super().__init__()
+        self.op = op
+        self.types = types
+        self.immediate_args = immediate_args if immediate_args is not None else []
+        self.args = args if args is not None else []
+
+        self.output_slots = [ScratchSlot() for _ in self.types]
+        self.reducer_set = False
+
+    def outputReducer(self, reducer: Callable[..., Expr]) -> Expr:
+        self.reducer = reducer
+        self.reducer_set = True
+        return self
```

Comment:
Have you considered an alternative implementation like this?

```python
def outputReducer(self, reducer: Callable[..., Expr]) -> Expr:
    # taken from __teal__
    input = [
        slot.load(self.types[i]) for i, slot in enumerate(self.output_slots)
    ]
    return Seq(self, reducer(*input))
```

It seems possibly simpler and more pure, since calling `outputReducer` does not modify the `MultiValue` object, plus the `MultiValue` is no longer responsible for converting the reducer function output to the IR. But I have not thought through all the implications.

Reply:
I agree, this approach is better.

Reply:
One issue I thought of. If someone does decide to call outputReducer() on MaybeValue output then the result won't be a MaybeValue, so hasValue() and value() won't be able to be called. I don't think this is a problem because if outputReducer() is called then it's because the user wants to do something with the return values different from the default. Another consideration is that reducers won't be able to be chained but we never intended for that to be supported anyway so imo not an issue. Any thoughts?